### PR TITLE
Enumeration of TypeIDs for both C and C++ code

### PR DESCRIPTION
**Pull request by abinashmeher999**
Addresses #517.

### CODE REVIEW

**Comment by isuruf** [comment]
Can you try using the same enum for both C & C++?

**Comment by abinashmeher999** [comment]
Do you mean without the `typedef`?

**Comment by isuruf** [comment]
Yes

**Comment by abinashmeher999** [comment]
Removed the `typedef` and used `enum TypeID` as the return type. Got this
```
/home/abinashmeher999/gsoc/symengine/symengine/cwrapper.h:79:6: error: use of enum ‘TypeID’ without previous declaration
 enum TypeID basic_get_type(const basic s);
      ^
```

**Comment by isuruf** [comment]
You don't need the enum keyword there

**Comment by abinashmeher999** [comment]
The same
```
/home/abinashmeher999/gsoc/symengine/symengine/cwrapper.h:79:1: error: ‘TypeID’ does not name a type
 TypeID basic_get_type(const basic s);
 ^
```

**Comment by abinashmeher999** [comment]
@certik @isuruf This is ready for review.

File: symengine/tests/cwrapper/test_cwrapper.c [hunk 1]
```diff
@@ -222,6 +222,20 @@ void test_free_symbols()
...
 
+void test_get_type() {
+    basic x, y;
+    basic_init(x);
+    basic_init(y);
+    symbol_set(x, "x");
+    integer_set_ui(y, 123);
+
+    SYMENGINE_C_ASSERT(basic_get_type(x)==SYMBOL);
+    SYMENGINE_C_ASSERT(basic_get_type(y)==INTEGER);
```
Comment (certik) [inline symengine/tests/cwrapper/test_cwrapper.c]: Add spaces around `==`.

Comment (abinashmeher999) [inline symengine/tests/cwrapper/test_cwrapper.c]: @certik done. Please check the new commit, as I squashed it to this one.

**Comment by certik** [comment]
So the code looks very clean and nice. But I don't like that the C++ version has the symbol `TypeID` (as well as all the enums) defined in a global namespace. It needs to go under a namespace.

I also don't like that the C interface does not have `SYMENGINE_` prefix in front. Without the prefix the symbols can clash with other libraries or user code.

For that reason, I think the way to go is the use the approach here: http://stackoverflow.com/questions/20792121/wrap-c-namespaced-enum-to-c-interface/20792525#20792525, as I mentioned a few times. Then symbols will be prefixed in C, thus no clashes there with user code, and in C++, the enum is now inside the SymEngine namespace, thus also no clashes there. This approach might require one cast, but I think that's fine, as the enum is only defined once, so things should still be safe.

**Comment by abinashmeher999** [comment]
I was thinking of creating a new PR with the approach you mentioned. I personally like that approach too. I was trying to finish this just to know if it works.

And regarding keeping it under the namespace, that can be done by the `#define __cplusplus` way. But it doesn't matter now. I am creating another PR with this approach.

**Comment by certik** [comment]
Sure, you can create a new PR.

Yes, you can put it into a C++ namespace with ifdef, but you still have to prefix it in C. So that's why we'll have to use that macro trick.

**Comment by certik** [comment]
Just keep this PR open to see if tests run.

**Comment by abinashmeher999** [comment]
OK

**Comment by certik** [comment]
Closing in favor of #519.

**Pull request closed without merging**